packages/block-library/src/gallery/edit.js

```diff
@@ -166,59 +115,33 @@ class GalleryEdit extends Component {
 		};
 	}
 
-	selectCaption( newImage, images, attachmentCaptions ) {
-		// The image id in both the images and attachmentCaptions arrays is a
-		// string, so ensure comparison works correctly by converting the
-		// newImage.id to a string.
-		const newImageId = toString( newImage.id );
-		const currentImage = find( images, { id: newImageId } );
-
-		const currentImageCaption = currentImage
-			? currentImage.caption
-			: newImage.caption;
-
-		if ( ! attachmentCaptions ) {
-			return currentImageCaption;
-		}
-
-		const attachment = find( attachmentCaptions, {
-			id: newImageId,
-		} );
-
-		// if the attachment caption is updated
-		if ( attachment && attachment.caption !== newImage.caption ) {
-			return newImage.caption;
-		}
-
-		return currentImageCaption;
-	}
-
 	onSelectImages( newImages ) {
-		const { columns, images, sizeSlug } = this.props.attributes;
-		const { attachmentCaptions } = this.state;
-		this.setState( {
-			attachmentCaptions: newImages.map( ( newImage ) => ( {
-				// Store the attachmentCaption id as a string for consistency
-				// with the type of the id in the images attribute.
-				id: toString( newImage.id ),
-				caption: newImage.caption,
-			} ) ),
+		const { clientId, replaceInnerBlocks } = this.props;
+		const { columns, sizeSlug, linkTo } = this.props.attributes;
+
+		const newBlocks = newImages.map( ( image ) => {
+			return createBlock( 'core/image', {
+				id: image.id,
+				caption: image.caption,
+				url: image.url,
+				link: image.link,
+				linkDestination: linkTo,
+				alt: image.alt,
+			} );
 		} );
+
 		this.setAttributes( {
 			images: newImages.map( ( newImage ) => ( {
 				...pickRelevantMediaFiles( newImage, sizeSlug ),
-				caption: this.selectCaption(
-					newImage,
-					images,
-					attachmentCaptions
-				),
 				// The id value is stored in a data attribute, so when the
 				// block is parsed it's converted to a string. Converting
 				// to a string here ensures it's type is consistent.
 				id: toString( newImage.id ),
 			} ) ),
 			columns: columns ? Math.min( newImages.length, columns ) : columns,
 		} );
+
+		replaceInnerBlocks( clientId, newBlocks );
```

Comment:
This may need to be replaced with getBlocksByClientId and insertBlocks to compare what is already in the gallery with what is passed back by media dialog if we want to preserve sort order changes between updates from the media browser